### DISCUSSION TITLE
Project's standards improvements proposal

### DISCUSSION THREAD

**Discussion by o-nnerb** [hmlongco/Factory#124] [Ideas]
I was thinking after the #113 discussion that some improvements can be done in this repository. Also, I like to make clear that I'm sharing this so we can discuss and collaborate for the proposed idea here.

1. Only main branch

    Using the main only branch to keep project commits, a general standard used by a lot of swift repositories, actually makes clear to check each PRs merged just by looking the commits historic.

2. Github Actions

    Adding and improving the implementation of some workflows for Github Actions so the PR is validated right from start. We can add severals workflows.
    - Test runner (also integrate with codecov.io)
    - Build validation for Apple and Linux (resolving the #113)
    - DocC runner for Github Pages.
  
3. Readme updates

    Something that I really miss here is the badges for Swift Package Index, they make clear to understand the current project compatibility and Swift version supports.
    
4. Adds templates

    Templates help us to contribute following some patterns during the write of an issue or PR description.

**Comment by hmlongco** [comment]
Appreciate the suggestions. Some thoughts...

1. If PRs exist they tend to exist on on main. I tend to use the develop branch for more extensive features and/or changes that are more in-depth and for which I'm still getting comments and input. That said, those changes could be PR'd to main.
2. Not a GitHub expert, so unsure what creating actions would entail. 
3. Readme... submit a PR?
4. If you have a suggestion for templates I'm open. 

But all of that said, this isn't a huge project where I'm constantly inundated with issues and PRs. 

The vast majority of the 30 PRs I've received, in fact, consist mostly of people fixing typos and spelling errors in the documentation. I think about two single line code changes have been merged into the repository. And I think about three or four other changes have been requested and either been declined or implemented in some other fashion.

When 2.2.0 is finalizedI think Factory will be pretty stable, unless something else comes up that I need to fix for. Xcode 15.

**Reply by o-nnerb** [reply]
Great Michael! My intention was to share my thoughts and you we’re very welcoming. 

- PRs

    You're right. With that in mind, actually, having main branch only with PR review could be unnecessary if you’re the only active contributor. I also agree about the features and stability of Factory, which just increase how unnecessary having main branch only might be.

- README

    I'll updated it. I was a little timid to do it without discussing before.

- Templates
 
    If I have any concrete suggestion about this, I'll open a new thread just for it.

**Reply by o-nnerb** [reply]
GitHub Actions can be implemented easily by creating the workflows at `.github/workflows/**.yml`

The GitHub documentation is very helpful about the job configuration and the steps. There's also some templates for beginners.

You can start with the `swift test` and increment it to run on different OS, like Linux. It’s the same pattern to also adds multiple builds for different versions of Swift.

The docc suggestion that I made is actually not needed. Swift Package Index do that for us.

The code coverage is also optional, but requires more advanced commands on macOS.

The last thing that I consider personal is that adding the workflow without the PR validation can be very painful and overwhelming. I have done this is the past which in some cases I released a version with build failing 🤡. I know, just check before releasing it 🤣

So I'll include this as an unnecessary feature base on the discussion in this thread.

**Comment by o-nnerb** [comment]
You can also check the build compilation status for Linux at Swift Package Index. 😉
